### PR TITLE
storage scrubber: GC ancestor shard layers

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

After a shard split, the pageserver leaves the ancestor shard's content in place.  It may be referenced by child shards, but eventually child shards will de-reference most ancestor layers as they write their own data and do GC.  We would like to eventually clean up those ancestor layers to reclaim space.

## Summary of changes

- Extend the physical GC command with `--mode=full`, which includes cleaning up unreferenced ancestor shard layers
- Add test `test_scrubber_physical_gc_ancestors`
- Remove colored log output: in testing this is irritating ANSI code spam in logs, and in interactive use doesn't add much.
- Refactor storage controller API client code out of storcon_client into a `storage_controller/client` crate
- During physical GC of ancestors, call into the storage controller to check that the latest shards seen in S3 reflect the latest state of the tenant, and there is no shard split in progress.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 1]
```diff
@@ -145,6 +263,122 @@ async fn maybe_delete_index(
...
+            //     },
+            //     Some(ttid_refs) => {
+
+            //     tracing::debug!("references for this ttid:");
+            //     for (layer_name, shard_index) in ttid_refs {
+            //         tracing::debug!("{}: {}", layer_name, shard_index);
+            //     }
+            //     }
+
+            // }
```
Comment (problame) [inline storage_scrubber/src/pageserver_physical_gc.rs:335]: Commented out code tends to rot. I'm ok with `if false { ... }`

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 2]
```diff
@@ -92,7 +163,7 @@ async fn maybe_delete_index(
             None => {
                 tracing::warn!("Missing last_modified");
                 summary.remote_storage_errors += 1;
-                return;
+                return false;
             }
             Some(last_modified) => {
                 let last_modified =
```
Comment (problame) [inline storage_scrubber/src/pageserver_physical_gc.rs:169]: The S3 SDK implements a

```
impl TryFrom<DateTime> for SystemTime {
```

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 3]
```diff
@@ -101,23 +172,70 @@ async fn maybe_delete_index(
...
         }
     };
-    if &age < min_age {
+    let old_enough = &age > min_age;
+
+    if !old_enough {
         tracing::info!(
             "Skipping young object {} < {}",
             age.as_secs_f64(),
             min_age.as_secs_f64()
```
Comment (problame) [inline storage_scrubber/src/pageserver_physical_gc.rs:192]: I think sprinkling in some `humantime::Duration` would be appreciated by whatever poor soul will have to read these logs?

**Review by problame — COMMENTED** [review]
We should clean up the variable names when speaking about tenants (`tenant_shard_id` vs `tenant_id`).

---


I would prefer we call it parent shard, so at least there is a minimal distinction from the word ancestor which we already use for `Timeline`.

---

What if the remote storage lists some but not all shards, e.g. because we're in the middle of `Tenant::split_prepare` ?  

Edit: ah, nice, you handled that.

Kind of an important thing to regress-test. Shouldn't be too hard to test this with a pausable failpoint, should it? 


Edit 2: but, did you also consider the same condition for the `stream_tenant_timelines` call made by `gc_ancestor`?  I think it can observe a tenant shard in S3 but some of its timelines' `IndexPart`s haven't been uploaded by `Tenant::split_prepare` yet.


Let's make sure we understand which races can happen and why they are fine / how they are handled. And: where will we document this? module-level comment doesn't seem like the worst choice, but, hard to discover.

---


About `AncestorRefs`: why is it not a more phyiscal refcount map, like on the `absolute_key` ? Sure, a bit less memory efficient but much harder to screw up. Similarly, I don't think we should go through the trouble of only tracking the cross-ancestor references.

I'm feeling kind of strongly about this, I think the code could be much simpler and lower risk.

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 4]
```diff
@@ -206,6 +444,8 @@ pub async fn pageserver_physical_gc(
             }
```
[the remark this answers is not in the document]
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs:444]: That's intended to be covered by the check in `into_gc_ancestors` that checks that we have data for the expected number of shards (`latest_shards.len() != latest_count.count()`)

Although I suppose there is an edge case where we have split more than once (e.g. 1,2,4), and we have errors loading all the latest shards, we would incorrectly conclude that the shard_count=2 ones are the latest.

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 3 above)
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs:192]: Pourquoi non.  6be4b75ce

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 1 above)
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs]: Left this by accident, removed in 8a3199519

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 5]
```diff
@@ -222,17 +462,51 @@ pub async fn pageserver_physical_gc(
...
+                &accumulator,
+            )
+        });
+        let mut timelines = std::pin::pin!(timelines.try_buffered(CONCURRENCY));
+
+        while let Some(i) = timelines.next().await {
+            let tl_summary = i?;
+
+            summary.indices_deleted += tl_summary.indices_deleted;
+            summary.remote_storage_errors += tl_summary.remote_storage_errors;
```
[the remark this answers is not in the document]
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs]: That counter was never nonzero here, but yeah, this was sloppy.  e125247fb

File: test_runner/regress/test_storage_scrubber.py [hunk 6]
```diff
@@ -111,6 +112,14 @@ def test_scrubber_tenant_snapshot(neon_env_builder: NeonEnvBuilder, shard_count:
     workload.validate()
 
 
+def drop_local_state(env, tenant_id):
+    env.storage_controller.tenant_policy_update(tenant_id, {"placement": "Detached"})
```
[the remark this answers is not in the document]
Comment (jcsp) [reply test_runner/regress/test_storage_scrubber.py:117]: b17a34d8a

File: test_runner/regress/test_storage_scrubber.py [hunk 7]
```diff
@@ -158,3 +163,99 @@ def test_scrubber_physical_gc(neon_env_builder: NeonEnvBuilder, shard_count: Opt
...
+    workload.churn_rows(100)
+    workload.churn_rows(100)
+    workload.churn_rows(100)
+    for shard in shards:
+        ps = env.get_tenant_pageserver(shard)
+        ps.http_client().timeline_compact(shard, timeline_id)
+        ps.http_client().timeline_gc(shard, timeline_id, 0)
+
+    # We will use a 1s threshold for deletion, let it pass
+    time.sleep(2)
```
[the remark this answers is not in the document]
Comment (jcsp) [reply test_runner/regress/test_storage_scrubber.py]: b17a34d8a

**Comment by jcsp** [comment]
> What if the remote storage lists some but not all shards, e.g. because we're in the middle of Tenant::split_prepare ?

I need to think about this more.  The hardest case is if we list things to do GC, then a new split happens while we are GC'ing, which references layers we will delete.  This is only an issue if doing multiple splits.

One possible soft approach is to always upload parent index_part during split, and then apply the min_age to parent index mtime

File: storage_scrubber/src/pageserver_physical_gc.rs (hunk 2 above)
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs]: ah, nice.  Used in 8c63696e0

**Comment by jcsp** [comment]
>. The hardest case is if we list things to do GC, then a new split happens while we are GC'ing, which references layers we will delete. This is only an issue if doing multiple splits.

Thought more about this.   When we GC an ancestor layer, it means that the indices for children no longer refer to it.  So any further splits cannot refer to the ancestral layer.

Importantly, the split process shuts down the remote timeline client for the parent shard before it uploads child shards.  That means that once we upload the child shards, the parents may be considered "frozen" and do not risk derefencing any ancestor layers (as long as the pageserver is still running).  If the pageserver dies halfway through a split, the storage controller will retry the split, which will eventually still write child shard indices which do not reference any ancestral layers that their parent doesn't reference.

**Comment by jcsp** [comment]
> About AncestorRefs: why is it not a more phyiscal refcount map, like on the absolute_key ? Sure, a bit less memory efficient but much harder to screw up.

Memory efficiency is relatively important here, because we build the AncestorRefs for all timelines in an entire region.

> Similarly, I don't think we should go through the trouble of only tracking the cross-ancestor references.

Same as above: because we're building this map globally, it's important that we limit the scope of what we store in memory.

Could we rework this to avoid building the monolithic ancestor refs map?  Yes, but at the cost of having to first build global list of tenant shards, then go back and do the per-tenant work (with a smaller AncestorRefs scoped to one tenant).  It's a tradeoff: for this PR, I'm preferring to take the memory cost of listing all ancestor refs, vs. the memory cost of storing a total list of all tenants (100ks).

I wrote this avoiding the assumption that listings are reliably returned in order, S3 kinda-sorta promises this, but I have limited trust in third party implementations which might offer a more readdir-like lack of ordering.

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 8]
```diff
@@ -26,59 +31,125 @@ pub enum GcMode {
...
+            GcMode::Full => write!(f, "full"),
         }
     }
 }
 
-async fn maybe_delete_index(
+// Map of cross-shard layer references, giving a refcount for each layer in each shard that is referenced by some other
+// shard in the same tenant.  This is sparse!  The vast majority of timelines will have no cross-shard refs, and those that
+// do have cross shard refs should eventually drop most of them via compaction.
+type AncestorRefs = BTreeMap<TenantTimelineId, HashMap<(ShardIndex, LayerName), usize>>;
```
[the remark this answers is not in the document]
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs]: Yeah, it makes it easier to read with comments to explain which things ref an ancestor vs. a descdendent: 232c7d4d2

**Comment by jcsp** [comment]
>I would prefer we call it parent shard, so at least there is a minimal distinction from the word ancestor which we already use for Timeline.

The use of "ancestor" is intentional, because a shard can reference layers that are from it's parent's parents.  It's annoying that we have an "ancestor shard" and "ancestor timeline" concept, but I think it's the right word.  We could call them "forebears" or some other synonym but I think that would be worse :-)

**Comment by problame** [comment]
> Could we rework this to avoid building the monolithic ancestor refs map? Yes, but at the cost of having to first build global list of tenant shards, then go back and do the per-tenant work (with a smaller AncestorRefs scoped to one tenant)


If we're going to [integrate scrubber with storcon anyway](https://neondb.slack.com/archives/C033RQ5SPDH/p1720688929093819?thread_ts=1720684348.341969&cid=C033RQ5SPDH) then we might as well avoid all the "discover tenants/shards through ListObjectsv2" and instead simply ask storcon for a list of tenants to scrub. And that list would include a consistent list of which shards exist.

**Review by problame — APPROVED** [review]
LGTM, sorry for the delay

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 9]
```diff
@@ -222,17 +536,61 @@ pub async fn pageserver_physical_gc(
...
+                &s3_client,
+                &bucket_config,
+                &min_age,
+                &target,
+                mode,
+                ttid,
+                &accumulator,
+            )
+        });
+        let mut timelines = std::pin::pin!(timelines.try_buffered(CONCURRENCY));
```
[the remark this answers is not in the document]
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs:555]: Yes, there's concurrency listing timelines, and then also concurrency analyzing them.

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 10]
```diff
@@ -26,20 +54,235 @@ pub enum GcMode {
...
+                    continue;
+                }
+                Ok(desc) => {
+                    // We expect to see that the latest shard count matches the one we saw in S3, and that none
+                    // of the shards indicate splitting in progress.
+                    if desc.shards.len() != latest_count.0 as usize {
+                        tracing::info!(%tenant_id, "Shard count seen in S3 ({latest_count:?}) doesn't match controller state ({})",
+                        desc.shards.len());
+                        continue;
+                    }
```
[the remark this answers is not in the document]
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs]: Good suggestion, modified in https://github.com/neondatabase/neon/pull/8196/commits/fde87d210246e21a6a103a92f1ee690e0c2c2ed3

File: storage_scrubber/src/pageserver_physical_gc.rs [hunk 11]
```diff
@@ -26,20 +54,235 @@ pub enum GcMode {
...
 
+async fn is_old_enough(
+    s3_client: &Client,
+    bucket_config: &BucketConfig,
+    min_age: &Duration,
+    key: &str,
+    summary: &mut GcSummary,
+) -> bool {
+    // Validation: we will only delete indices after one week, so that during incidents we will have
+    // easy access to recent indices.
```
[the remark this answers is not in the document]
Comment (jcsp) [reply storage_scrubber/src/pageserver_physical_gc.rs]: updated in https://github.com/neondatabase/neon/pull/8196/commits/fde87d210246e21a6a103a92f1ee690e0c2c2ed3

**Review by arssher — APPROVED** [review]

**Pull request merged**